server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder2x.java

```diff
@@ -79,21 +85,42 @@ public boolean readHeader(ByteBuf buffer, byte version, long messageId, HotRodHe
          }
          buffer.markReaderIndex();
       }
-      int flag = ExtendedByteBufJava.readMaybeVInt(buffer);
-      if (flag == Integer.MIN_VALUE) {
-         return false;
+
+      if (header.flag == null) {
```

Comment:
If you want to use the if statements to check if it is null, then you have to mark the read buffer at the end of the if block.  Otherwise a retry could get messed up.  This is true for all 3.